### PR TITLE
Linked Expenses with Suppliers

### PR DESCRIPTION
**Pull request by emi-silva**
This pull request corresponds to the third implementation step from Epic #2088.

### CODE REVIEW

File: application/views/expenses/form.php [hunk 1]
```diff
@@ -201,13 +251,16 @@
...
+		ignore: '',
+
 		rules:
 		{
 			category: 'required',
 			date:
 			{
 				required: true
 			},
+			supplier_id: 'required',
```
[the remark this answers is not in the document]
Comment (emi-silva) [reply application/views/expenses/form.php]: I would say so, what do you think?

File: application/migrations/20180819195000_expenses_supplier_id.php [hunk 2]
```diff
@@ -0,0 +1,97 @@
+<?php if (!defined('BASEPATH')) exit('No direct script access allowed');
```
Comment (daN4cat) [inline application/migrations/20180819195000_expenses_supplier_id.php]: Please use the SQL upgrade script, I moved your script to SQL only from the previous commit as a courtesy but I would like to avoid to do it every time.

At least for vs 3.3.0 we keep that way, for future versions we will do differently and probably using a feature based as long as we don't have inter dependencies.

The rationale for this change is that if we touch the SQL date we need to align this too to avoid this upgrade to start before the other and fail.

File: application/views/expenses/form.php [hunk 3]
```diff
@@ -187,6 +197,46 @@
...
+
+	$('#remove_supplier_button').click(function() {
+		$('#supplier_id').val('');
+		$('#supplier_name').removeAttr('readonly');
+		$('#supplier_name').val('');
+		$(this).css('display', 'none');
+	});
+
+	<?php
+		if(!empty($expenses_info->expense_id)) {
```
Comment (daN4cat) [inline application/views/expenses/form.php]: no reason for this extra TAB in the whole section.

File: application/views/expenses/form.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (daN4cat) [reply application/views/expenses/form.php]: Ehm... I would prefer not having it mandatory because some people might just want to capture some petty cash transaction without being a supplier.
Let say you refund an employee of an extra expense or similar.

File: public/css/bootstrap.autocomplete.css [hunk 4]
```diff
@@ -3,7 +3,7 @@
     position: absolute;
     top: 100%;
     left: 0;
-    z-index: 1000;
+    z-index: 1050;
```
Comment (daN4cat) [inline public/css/bootstrap.autocomplete.css]: what's the purpose of this change?

File: application/views/expenses/form.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (emi-silva) [reply application/views/expenses/form.php]: OK

File: application/migrations/20180819195000_expenses_supplier_id.php (hunk 2 above)
Comment (emi-silva) [reply application/migrations/20180819195000_expenses_supplier_id.php]: OK

File: application/views/expenses/form.php (hunk 3 above)
Comment (emi-silva) [reply application/views/expenses/form.php]: OK

File: public/css/bootstrap.autocomplete.css (hunk 4 above)
Comment (emi-silva) [reply public/css/bootstrap.autocomplete.css]: This is necessary so that the dropdown that lets you choose the supplier appears in front of the New Expense form and not behind.

**Comment by emi-silva** [comment]
@daN4cat and @jekkos: Please check my new commit. Thanks!

File: application/controllers/Expenses.php [hunk 5]
```diff
@@ -139,6 +139,10 @@ public function save($expense_id = -1)
 			'deleted' => $this->input->post('deleted') != NULL
 		);
 
+        if($expense_data['supplier_id'] == '') {
```
Comment (jekkos) [inline application/controllers/Expenses.php]: check indentation, it's not perfectly aligned

File: application/migrations/sqlscripts/3.2.1_to_3.3.0.sql [hunk 6]
```diff
@@ -111,3 +111,36 @@ ALTER TABLE `ospos_suppliers`
...
+
+
+--
+-- Link Expenses with Suppliers
+--
+
+-- Add supplier id
+
+ALTER TABLE `ospos_expenses`
+  ADD COLUMN `supplier_id` int(10) NULL;
```
Comment (jekkos) [inline application/migrations/sqlscripts/3.2.1_to_3.3.0.sql]: do we have the same datatype as the id column in suppliers table?

File: application/migrations/sqlscripts/3.2.1_to_3.3.0.sql [hunk 7]
```diff
@@ -111,3 +111,36 @@ ALTER TABLE `ospos_suppliers`
...
+
+-- Link suppliers
+
+UPDATE `ospos_expenses`
+  INNER JOIN `ospos_suppliers`
+    ON `ospos_expenses`.`supplier_name` = `ospos_suppliers`.`company_name`
+SET `ospos_expenses`.`supplier_id` = `ospos_suppliers`.`person_id`;
+
+-- Save name in description for those expenses whose supplier isn't registered
+
```
Comment (jekkos) [inline application/migrations/sqlscripts/3.2.1_to_3.3.0.sql]: @daN4cat was the idea to stick to using 3.2.0 to 3.3.0 sql for this version only?

File: application/views/expenses/form.php [hunk 8]
```diff
@@ -228,14 +228,14 @@
...
 	<?php
-		if(!empty($expenses_info->expense_id)) {
+	if(!empty($expenses_info->expense_id)) {
 	?>
-			$('#supplier_id').val('<?php echo $expenses_info->supplier_id ?>');
-			$('#supplier_name').val('<?php echo $expenses_info->supplier_name ?>');
-			$('#supplier_name').attr('readonly', 'readonly');
-			$('#remove_supplier_button').css('display', 'inline-block');
+		$('#supplier_id').val('<?php echo $expenses_info->supplier_id ?>');
+		$('#supplier_name').val('<?php echo $expenses_info->supplier_name ?>');
```
Comment (jekkos) [inline application/views/expenses/form.php]: I think you can chain both calls like attr().val()

**Comment by jekkos** [comment]
Looks good, just some minor remakrs

File: application/migrations/sqlscripts/3.2.1_to_3.3.0.sql (hunk 6 above)
Comment (emi-silva) [reply application/migrations/sqlscripts/3.2.1_to_3.3.0.sql]: Yes. You can check this with the following SQL command:

    > show columns from ospos_suppliers where Field = 'person_id';
    +-----------+---------+------+-----+---------+-------+
    | Field     | Type    | Null | Key | Default | Extra |
    +-----------+---------+------+-----+---------+-------+
    | person_id | int(10) | NO   | MUL | NULL    |       |
    +-----------+---------+------+-----+---------+-------+
    1 row in set (0.00 sec)

File: application/views/expenses/form.php (hunk 8 above)
Comment (emi-silva) [reply application/views/expenses/form.php]: OK

**Comment by emi-silva** [comment]
@jekkos I just added a commit with your requested changes. Thank you very much.

File: application/controllers/Expenses.php [hunk 9]
```diff
@@ -139,6 +139,10 @@ public function save($expense_id = -1)
 			'deleted' => $this->input->post('deleted') != NULL
 		);
 
+		if($expense_data['supplier_id'] == '') {
```
Comment (daN4cat) [inline application/controllers/Expenses.php]: We try to be consistent with the rest of the code, even if it's an annoying detail...

Comment (emi-silva) [reply application/controllers/Expenses.php]: OK, sorry about that...

**Comment by emi-silva** [comment]
@daN4cat @jekkos please review my last commit. I think we're ready to merge? I can squash the commits if necessary

**Comment by daN4cat** [comment]
Looks fine, I will merge the code in the next days. First I need to test and then merge.

Thank you for the contribution.

**Comment by emi-silva** [comment]
You're very welcome. Thank you for your reviews and for testing and merging and everything you do to make OSPOS work.

**Comment by emi-silva** [comment]
@daN4cat I can't reproduce the error you mention. I pulled your commits and tested it, and it works fine for me. Then I cloned a new repository, created a new database, then created a supplier, and then in Expenses the suggestion dropdown showed it correctly. Please show me the steps to reproduce this error so I can correct it. On Sunday I'm leaving for a long trip through South America and won't be able to do any coding for a few months.

**Comment by SteveIreland** [comment]
Should this pull request be closed?
